### PR TITLE
Solution

### PR DESCRIPTION
**Pull request by JankaGramofonomanka**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,5 +1,88 @@
...
 
 ## About
-This is my solution to a recruitment task
+This is my solution to a recruitment task, which is described in 
+`task_description.pdf`
+
+The application will be implemented with accordance to the API defined in 
+`touk-task-api.yaml`
+
+## Prerequisites
```
Comment (arkadius) [inline README.md:10]: Nice description :+1:

File: conf/application.conf [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# https://www.playframework.com/documentation/latest/Configuration
+play.modules.enabled += "play.modules.reactivemongo.ReactiveMongoModule"
+mongodb.uri = "mongodb+srv://server:NH89EQS8Sjc9ESYS@cluster0.lz8tw.mongodb.net/multiplex-example-db"
```
Comment (arkadius) [inline conf/application.conf:3]: Where can I found this db? Can I run this service locally with e.g. mongodb run in docker?

File: app/views/main.scala.html [hunk 3]
```diff
@@ -0,0 +1,25 @@
+@*
```
Comment (arkadius) [inline app/views/main.scala.html:1]: Can you remove files and related code that just only obfuscating this service?

File: app/lib/Server.scala [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+
+
+  // Request body processing ----------------------------------------------------------------------
+  def processReservationData(
+    screeningId:  ScreeningId,
+    body:         JsValue,
+  ): Option[Reservation] = body match {
+
+    case JsObject(obj) => for {
+      jsonName    <- obj.get("name")
```
Comment (arkadius) [inline app/lib/Server.scala:61]: Don't you know some library that will handle this json extraction by you so you'll need only declare some case class. What if you want to use the same structure in two places - how you ensure if you don't do some typo in one place?

File: app/lib/Server.scala [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+  def serveListScreeningsRequest(
+    date: String,
+    from: String,
+    to:   String,
+  ): EitherT[Future, Error, JsValue] = {
+
+    val dateFrom  = s"$date $from"
+    val dateTo    = s"$date $to"
+
+    val tryFrom = Try(DateTimeFormat.forPattern("dd-MM-yyyy HH:mm").parseDateTime(dateFrom))
```
Comment (arkadius) [inline app/lib/Server.scala:145]: Why such date format? It looks like mixture of polish date format and iso date format

File: app/lib/Server.scala [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+
+    info        <- dbInterface.getScreeningInfo(reservation.screening)
+    takenSeats  <- getTakenSeats(reservation.screening, info.room)
+
+    requestedSeats = reservation.seats.keys.toList
+
+    // Not to late to reserve seats
+    expirationDate = info.start - 15.minutes
+    _ <- EitherT.cond[Future](
+      DateTime.now() < expirationDate,
```
Comment (arkadius) [inline app/lib/Server.scala:179]: Do you know how to change this code in order to make possible to write test scenarios with different combinations of reservation time is after / before expiration date?

File: app/lib/Server.scala [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+      + ("rows" -> Json.toJson(availableSeats.dim.numRows))
+      + ("seats-per-row" -> Json.toJson(availableSeats.dim.numColumns))
+      + ("available-seats" -> Json.toJson(availableSeats.seats))
+    )
+
+  } yield result
+
+
+  // Request body processing ----------------------------------------------------------------------
+  def processReservationData(
```
Comment (arkadius) [inline app/lib/Server.scala:55]: You don't think that it would be a good idea to write unit tests for this method? How you will know that you handle all edge cases?

File: app/controllers/HomeController.scala [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  /**
+   * Create an Action to render an HTML page.
+   *
+   * The configuration in the `routes` file means that this method
+   * will be called when the application receives a `GET` request with
+   * a path of `/`.
+   */
+  def index() = Action { implicit request: Request[AnyContent] =>
+    Ok(views.html.index())
+  }
```
Comment (gadomsky) [inline app/controllers/HomeController.scala:37]: Do we need that?

File: app/lib/DBInit.scala [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+  // Process data ---------------------------------------------------------------------------------
+  def processData(data: JsValue): Option[InitialData] = data match {
+    case JsObject(obj) => for {
+      jsonDate        <- obj.get("date")
+      jsonRooms       <- obj.get("rooms")
+      jsonScreenings  <- obj.get("screenings")
+
+      dateStr <- jsonDate.asOpt[String]
+      date <- Try(DateTimeFormat.forPattern("dd-MM-yyyy").parseDateTime(dateStr)).toOption
```
Comment (gadomsky) [inline app/lib/DBInit.scala:50]: This way you loose information about what is wrong (e.g. wrong date format), and as a result you return generic error `"invalid request body"`

File: app/lib/DBInterface.scala [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+import com.github.nscala_time.time.Imports._
+
+import lib.Data._
+
+
+class DBInterface(api: ReactiveMongoApi) {
+
+  // Getting data from the database----------------------------------------------------------------
+  def getReservations(screeningId: ScreeningId): EitherT[Future, Error, List[Reservation]] = {
```
Comment (gadomsky) [inline app/lib/DBInterface.scala:24]: What do you think about extracting trait with all these methods, so we could have different implementations of repository (e.g. storing data in memory instead of real db)?

File: app/lib/DBInterface.scala [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+    val futureQueryResult: Future[List[BSONDocument]] = for {
+      db <- api.database
+      queryDoc = BSONDocument("screening-id" -> screeningId)
+      reservations <- db.collection[BSONCollection]("reservations")
+        .find(queryDoc)
+        .cursor[BSONDocument]()
+        .collect[List](1000, Cursor.FailOnError[List[BSONDocument]]())
```
Comment (gadomsky) [inline app/lib/DBInterface.scala:31]: 1000 - what is it? Is it some kind of limit of data? What if we have more? How do you think we can handle it?

File: app/lib/Data.scala [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+    seats:      Map[Seat, TicketType],
+    reserver:   Person,
+  )
+
+  // Utils ----------------------------------------------------------------------------------------
+  def ticketPrice(ticket: TicketType): Double = ticket match {
+    case Adult    => 25
+    case Student  => 18
+    case Child    => 12.5
+  }
```
Comment (gadomsky) [inline app/lib/Data.scala:73]: Do you know any way to pass this information but not hardcode it?

File: app/lib/Server.scala [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+  def validateReservation(
+    reservation: Reservation,
+  ): EitherT[Future, Error, DateTime] = for {
+
+    info        <- dbInterface.getScreeningInfo(reservation.screening)
+    takenSeats  <- getTakenSeats(reservation.screening, info.room)
+
+    requestedSeats = reservation.seats.keys.toList
+
+    // Not to late to reserve seats
```
Comment (gadomsky) [inline app/lib/Server.scala:176]: Maybe we could extract properly named methods instead of commenting each step?

**Review by gadomsky — COMMENTED** [review]
Thanks for the solution, I have left some comments. Futhermore:
 * I miss seperation of different parts in you're code. E.g. domain, infrastrucutre, utils etc.
 * I do not see any tests :( 
 * I don't see any private vals/defs. All yout methods and fields are public even if they do not need to be.
 * Handling bson/json everywhere makes hard to focus on "business code" when reading. Maybe there are some external libs which help with those things?
 * Demo scripts are nice :)
 * Plus for attaching openapi docs

File: app/lib/Server.scala (hunk 6 above)
Comment (JankaGramofonomanka) [reply app/lib/Server.scala:179]: I should have made the current time a parameter of `validateReservation` and `serveReservationRequest` and call `DateTime.now()` in `HomeController.postScreening`.

File: app/lib/Server.scala (hunk 5 above)
Comment (JankaGramofonomanka) [reply app/lib/Server.scala:145]: I didn't think what would be the best format, I just picked the first one that came to my mind.

File: app/controllers/HomeController.scala (hunk 8 above)
Comment (JankaGramofonomanka) [reply app/controllers/HomeController.scala:37]: I guess not, It's a leftover from the play-framework template.

File: app/lib/DBInterface.scala (hunk 10 above)
Comment (JankaGramofonomanka) [reply app/lib/DBInterface.scala:24]: I should have done it. I thought of it myself, unfortunately after I submitted the solution.

File: app/lib/DBInterface.scala (hunk 11 above)
Comment (JankaGramofonomanka) [reply app/lib/DBInterface.scala:31]: I meant to replace this value with some limit, that I would read from a config file or from an environment variabkle but I forgot about it. The reason there is a limit, is that I haven't found any other way to collect data then this `collect` method.

File: app/lib/Data.scala (hunk 12 above)
Comment (JankaGramofonomanka) [reply app/lib/Data.scala:73]: I could put it in a config file. In real life the prices would probably be stored in the database.

File: app/lib/Server.scala (hunk 13 above)
Comment (JankaGramofonomanka) [reply app/lib/Server.scala:176]: Yes.

File: app/lib/DBInit.scala (hunk 9 above)
Comment (JankaGramofonomanka) [reply app/lib/DBInit.scala:50]: You're right, I should have made the `processData` method return an `Either[Error, InitialData]`.

File: conf/application.conf (hunk 2 above)
Comment (JankaGramofonomanka) [reply conf/application.conf:3]: I just realized I only allowed my own IP address to access the database, it should now be accessible from anywhere.
Sorry for that. Nonetheless, as I wrote in the README, you can set up your own database and replace the uri. You can also run mongodb locally, then you should replace the uri with "mongodb://localhost:27017/\<DATABASE-NAME\>".

File: README.md (hunk 1 above)
Comment (JankaGramofonomanka) [reply README.md:10]: Thanks!

File: app/lib/Server.scala (hunk 4 above)
Comment (JankaGramofonomanka) [reply app/lib/Server.scala:61]: I don't know of any such library. Since I'm new to scala, I already had to find and learn all the other frameworks and I found it easier to write the json extraction myself, than to search for and figure aout another library.
If I wanted to use the same structure somewhere else, I would refactor this method so that I would't repeat the code or I would just create another module for processing json and put this method there.

File: app/views/main.scala.html (hunk 3 above)
Comment (JankaGramofonomanka) [reply app/views/main.scala.html:1]: I should have done that but I was occupied with other things.

File: app/lib/Server.scala (hunk 7 above)
Comment (JankaGramofonomanka) [reply app/lib/Server.scala:55]: It would, but I left testing for the end and then there was no time for it.

**Comment by JankaGramofonomanka** [comment]
> Thanks for the solution, I have left some comments. Futhermore:
> 
> * I miss seperation of different parts in you're code. E.g. domain, infrastrucutre, utils etc.

I have nothing for my defense here, I am not familiar with any good practices of separating code, I do it by intuition.

> * Handling bson/json everywhere makes hard to focus on "business code" when reading. Maybe there are some external libs which help with those things?

As I said in some other comment, I am new to Scala and I had to learn the other libraries while doing this task and I found it easier to handle json myself, then search for and learn another library.

> * I do not see any tests :(

For the same reason as above, to write tests, I would have to learn how it is done in Scala, so I left it for the end, and then there was no time. To be honest I always leave tests for the end and have little or no time for them.

> * I don't see any private vals/defs. All yout methods and fields are public even if they do not need to be.

I was generally writing this project as I would in Haskell and there are no private fields and methods in Haskell, so I wasn't thinking of them.

> * Demo scripts are nice :)
> * Plus for attaching openapi docs

2 x Thanks.
